src/nmodl/nocpout.cpp

```diff
@@ -847,20 +876,33 @@ static const char *_mechanism[] = {\n\
     }
 
     Lappendstr(defs_list,
-               "\n\
-extern Prop* need_memb(Symbol*);\n\n\
-static void nrn_alloc(Prop* _prop) {\n\
-	Prop *prop_ion;\n\
-	double *_p; Datum *_ppvar;\n\
-");
+               "\n"
+               "extern Prop* need_memb(Symbol*);\n"
+               "static void nrn_alloc(Prop* _prop) {\n"
+               "  Prop *prop_ion{};\n"
+               "  Datum *_ppvar{};\n");
     if (point_process) {
         Lappendstr(defs_list,
-                   " if (nrn_point_prop_) {\n\
-	_prop->_alloc_seq = nrn_point_prop_->_alloc_seq;\n\
-	_p = nrn_point_prop_->param;\n\
-	_ppvar = nrn_point_prop_->dparam;\n }else{\n");
+                   " if (nrn_point_prop_) {\n"
+                   "    _nrn_mechanism_access_alloc_seq(_prop) = "
+                   "_nrn_mechanism_access_alloc_seq(nrn_point_prop_);\n"
+                   "    _ppvar = _nrn_mechanism_access_dparam(nrn_point_prop_);\n"
+                   "  } else {\n");
     }
-    Sprintf(buf, "	_p = nrn_prop_data_alloc(_mechtype, %d, _prop);\n", parraycount);
+    // need to fill _prop->dparam before calling _nrn_mechanism_cache_range(Prop*)
+    if (ppvar_cnt) {
+        Sprintf(buf, "  _ppvar = nrn_prop_datum_alloc(_mechtype, %d, _prop);\n", ppvar_cnt);
+        Lappendstr(defs_list, buf);
+        Lappendstr(defs_list, "   _nrn_mechanism_access_dparam(_prop) = _ppvar;\n");
+    }
+    // seems that even in the old code and with vectorize == false that the global _p, _ppvar were
+    // shadowed, so don't worry about shadowing the global _ml and _iml here
+    Sprintf(buf,
+            "    _nrn_mechanism_cache_instance _ml_real{_prop};\n"
+            "    auto* const _ml = &_ml_real;\n"
```

Comment:
`note here for myself`: during the first of comparing generated cpp code (e.g. hh.cpp), it wasn't obvious the usage of `_ml_real`, `_ml` and `_ml_arg`.